src/cache.rs

```diff
@@ -6,128 +6,80 @@
+/// path is the path of the rocks db database
 pub(crate) struct Cache {
-    data: Map<PublicSpendKey, KeyData>,
+    db: DB,
+    write_batch: WriteBatch,
 }
 
 impl Cache {
-    /// Returns a new cache instance. Before this is called the cache data path
-    /// must be set with [`set_data_path`].
-    ///
-    /// # Panics
-    /// If called before [`set_data_path`].
-    pub(crate) fn new() -> Result<Self, Error> {
-        let data_dir = CACHE_DATA_PATH.get().expect("cache path must be set");
-        let id_path = data_dir.join("cache.id");
-
-        let persisted_id = match id_path.exists() {
-            true => {
-                let bytes = fs::read(id_path)?;
-                let mut source = Source::new(&bytes);
-
-                Some(PersistedId::decode(&mut source)?)
-            }
-            false => None,
-        };
-
-        let data = persisted_id.map_or(Ok(Map::new()), |id| id.restore())?;
-        Ok(Self { data })
-    }
-
-    /// Sets the cache data directory. This function can be called once and is a
-    /// requirement for being able to instantiate a cache using [`new`].
-    ///
-    /// # Panics
-    /// If called more than once or if unable to remove old cache file.
-    pub(crate) fn set_data_path(data_dir: PathBuf) -> Result<(), Error> {
-        // remove old cache if it's still there
-        let db_path = data_dir.join("cache.db");
-        if db_path.exists()
-            && !db_path.is_dir()
-            && fs::remove_file(&db_path).is_err()
-        {
-            panic!("Failed to remove old cache files. Please remove {} manually and try again.", db_path.display())
-        }
-
-        CACHE_DATA_PATH
-            .set(data_dir)
-            .expect("cache path can only be set once");
-
-        Persistence::with_backend(&backend(), |_| Ok(()))?;
-
-        Ok(())
-    }
-
-    /// Persist the cache to the internal backend.
-    pub(crate) fn persist(&self) -> Result<(), Error> {
-        let data_dir = CACHE_DATA_PATH.get().expect("cache path must be set");
-        let id_path = data_dir.join("cache.id");
-
-        let persistence_id = Persistence::persist(&backend(), &self.data)?;
-
-        let mut bytes = [0u8; 36];
-        let mut sink = Sink::new(&mut bytes);
-
-        persistence_id.encode(&mut sink);
+    /// Returns a new cache instance.
+    pub(crate) fn new<T: AsRef<Path>>(path: T) -> Result<Self, Error> {
+        let mut opts = Options::default();
+        opts.create_if_missing(true);
+        // After 10 million bytes, sort the cache file and create new one
+        opts.set_write_buffer_size(10_000_000);
 
-        fs::write(id_path, bytes)?;
+        let db = DB::open(&opts, path)?;
+        let write_batch = WriteBatch::default();
 
-        Ok(())
+        Ok(Self { db, write_batch })
     }
 
-    /// Insert a note into the cache at the given block height. If note is
-    /// `None` only the block height gets updated.
+    /// We store (key, Vec<(height, note)>).
```

Comment:
I'm for store (height, note) and use ColumnFamily to filter the key.
If this is not possible, I'm for store (key, height, note)